storage/range.go

```diff
@@ -279,44 +279,122 @@ func (r *Range) IsFirstRange() bool {
-			return &proto.NotLeaderError{}
-		}
+// redirectOnOrAcquireLeaderLease checks whether this replica has the
+// leader lease at the specified timestamp. If it does, returns
+// success. If another replica currently holds the lease, redirects by
+// returning NotLeaderError. If the lease is expired, a renewal is
+// synchronously requested. This method uses the leader lease mutex
+// to guarantee only one request to grant the lease is pending.
+//
+// TODO(spencer): implement threshold regrants to avoid latency in
+//  the presence of read or write pressure sufficiently close to the
+//  current lease's expiration.
+//
+// TODO(spencer): for write commands, don't wait while requesting
+//  the leader lease. If the lease acquisition fails, the write cmd
+//  will fail as well. If it succeeds, as is likely, then the write
+//  will not incur latency waiting for the command to complete.
+//  Reads, however, must wait.
+func (r *Range) redirectOnOrAcquireLeaderLease(args proto.Request) error {
+	// Skip check for a lease acquisition request.
+	if _, ok := args.(*proto.InternalLeaderLeaseRequest); ok {
+		return nil
 	}
-	if proto.IsReadOnly(args) {
-		if header.ReadConsistency == proto.CONSENSUS {
-			return util.Errorf("consensus reads not implemented")
-		} else if header.ReadConsistency == proto.INCONSISTENT && header.Txn != nil {
-			return util.Errorf("cannot allow inconsistent reads within a transaction")
+	r.llMu.Lock()
+	defer r.llMu.Unlock()
+	// If lease is currently held by another, redirect to holder.
+	timestamp := args.Header().Timestamp
+	if held, expired := r.HasLeaderLease(timestamp); !held && !expired {
+		return r.newNotLeaderError()
+	} else if !held || expired {
+		// Otherwise, if not held by this replica or expired, request renewal.
+		if err := r.requestLeaderLease(timestamp); err != nil {
+			return util.Errorf("failed to acquire lease on read pressure: %s", err)
 		}
 	}
-	if !r.ContainsKeyRange(header.Key, header.EndKey) {
-		return proto.NewRangeKeyMismatchError(header.Key, header.EndKey, r.Desc())
-	}
 	return nil
 }
 
+// verifyLeaderLease checks whether the requesting replica (by raft
+// node ID) holds the leader lease covering the specified timestamp.
+func (r *Range) verifyLeaderLease(originRaftNodeID multiraft.NodeID, timestamp proto.Timestamp) bool {
+	l := r.getLease()
+	return uint64(originRaftNodeID) == l.RaftNodeID && timestamp.Less(l.Expiration)
+}
+
+// HasLeaderLease returns whether this replica holds or was the last
+// holder of the leader lease, and whether the lease has expired.
+// Leases may not overlap, and a gap between successive lease holders
+// is expected.
+func (r *Range) HasLeaderLease(timestamp proto.Timestamp) (bool, bool) {
+	if l := r.getLease(); l.RaftNodeID != 0 {
+		held := l.RaftNodeID == uint64(r.rm.RaftNodeID())
+		var expired bool
+		if held {
+			// Held by us; compare to expiration.
+			expired = !timestamp.Less(l.Expiration)
```

Comment:
so if I'm trying to read at a timestamp that predates when the current lease holder acquired its lease that's fine, right?